From 4747deaa64b7b8d1d79bc50763929959942e4e59 Mon Sep 17 00:00:00 2001
From: Hauke Mehrtens <hauke.mehrtens@intel.com>
Date: Wed, 28 Oct 2020 16:58:01 +0100
Subject: [PATCH] PONRTSYS-8201: Use blocking notifier instead of atomic

The atomic notifier does a rcu_read_lock() around the
atomic_notifier_call_chain() which is needed when it is called from
atomic context. The blocking notifgier only uses a semaphore in
blocking_notifier_call_chain() which is not blocking other soft
interrupts. As long as DP_SPIN_LOCK is not set in DP this function is
not called in atomci context. Currently DP_SPIN_LOCK is not set.
---
 drivers/net/datapath/dpm/datapath.h          |  2 +-
 drivers/net/datapath/dpm/datapath_notifier.c | 19 ++++++++++++++-----
 2 files changed, 15 insertions(+), 6 deletions(-)

diff --git a/drivers/net/datapath/dpm/datapath.h b/drivers/net/datapath/dpm/datapath.h
index dd49e1693036..5a12d8957093 100644
--- a/drivers/net/datapath/dpm/datapath.h
+++ b/drivers/net/datapath/dpm/datapath.h
@@ -1053,7 +1053,7 @@ struct dp_evt_notif_data {
 	int inst;
 };
 
-extern struct atomic_notifier_head dp_evt_notif_list;
+extern struct blocking_notifier_head dp_evt_notif_list;
 int register_dp_event_notifier(struct dp_event *info);
 int unregister_dp_event_notifier(struct dp_event *info);
 
diff --git a/drivers/net/datapath/dpm/datapath_notifier.c b/drivers/net/datapath/dpm/datapath_notifier.c
index 9ace803efc14..18683441de62 100644
--- a/drivers/net/datapath/dpm/datapath_notifier.c
+++ b/drivers/net/datapath/dpm/datapath_notifier.c
@@ -63,7 +63,15 @@ int dp_notifier_invoke(int inst, struct net_device *dev,
 		return DP_FAILURE;
 	}
 
-	atomic_notifier_call_chain(&dp_evt_notif_list, type, &notif_data);
+#ifdef DP_SPIN_LOCK
+#warning blocking_notifier will not work with DP_SPIN_LOCK, use atomic_notifier
+#endif
+	/* blocking_notifier_call_chain uses a semaphore to lock the execution,
+	 * the atomic_notifier uses a rcu_read_lock. Using the semaphore is
+	 * preferred, but only possible when it is not called inside a spinlock.
+	 * When DP_SPIN_LOCK is defined this is called from a spinlock.
+	 */
+	blocking_notifier_call_chain(&dp_evt_notif_list, type, &notif_data);
 
 	return DP_SUCCESS;
 }
@@ -261,7 +269,7 @@ int dp_event(struct notifier_block *this, unsigned long event, void *ptr)
 	return 0;
 }
 
-ATOMIC_NOTIFIER_HEAD(dp_evt_notif_list);
+BLOCKING_NOTIFIER_HEAD(dp_evt_notif_list);
 static int dp_notifier_cb(struct notifier_block *self, unsigned long action,
 		       void *data)
 {
@@ -362,7 +370,8 @@ int register_dp_event_notifier(struct dp_event *info)
 		 * remove nb from list, modify evt_info and re-insert the nb
 		 */
 		nb = info->id;
-		ret = atomic_notifier_chain_unregister(&dp_evt_notif_list, nb);
+		ret = blocking_notifier_chain_unregister(&dp_evt_notif_list,
+							 nb);
 		if (ret) {
 			pr_err("%s: invalid handle\n", __func__);
 			return DP_FAILURE;
@@ -372,7 +381,7 @@ int register_dp_event_notifier(struct dp_event *info)
 	}
 
 	memcpy(&evt_notif->evt_info, info, sizeof(*info));
-	atomic_notifier_chain_register(&dp_evt_notif_list, nb);
+	blocking_notifier_chain_register(&dp_evt_notif_list, nb);
 	return 0;
 }
 
@@ -386,7 +395,7 @@ int unregister_dp_event_notifier(struct dp_event *info)
 		return DP_FAILURE;
 	}
 
-	ret = atomic_notifier_chain_unregister(&dp_evt_notif_list, info->id);
+	ret = blocking_notifier_chain_unregister(&dp_evt_notif_list, info->id);
 	if (ret) {
 		pr_err("%s:%d: invalid handle\n", __func__, __LINE__);
 		return DP_FAILURE;
